### PR TITLE
impl(generator): refactor DiscoveryTypeVertex::FormatProperties

### LINKED ISSUE #12234: Refactor DiscoveryVertexType::FormatProperties into smaller functions
Over time this function has grown and could use a refactoring pass to break it into more cohesive functions that are easier to grok.

### PR DESCRIPTION
**Pull request by scotthart**
fixes #12234

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243)

### CODE REVIEW

File: generator/internal/discovery_type_vertex.cc [hunk 1]
```diff
@@ -230,6 +258,32 @@ DiscoveryTypeVertex::DetermineReservedAndMaxFieldNumbers(
   return message_properties;
 }
 
+Status DiscoveryTypeVertex::UpdateTypeNames(
```
Comment (dbolduc) [inline generator/internal/discovery_type_vertex.cc:261]: nit: I cannot see why this is a class member vs. a standalone function. But surely the compiler would have yelled at you to make it `static`. I might just be blind.

**Review by dbolduc — APPROVED** [review]

File: generator/internal/discovery_type_vertex.cc [hunk 2]
```diff
@@ -44,6 +44,34 @@ absl::optional<std::string> CheckForScalarType(nlohmann::json const& j) {
...
+                                   std::string(field["enumDescriptions"][i]));
+      } else {
+        enum_comments.emplace_back(std::string(field["enum"][i]),
+                                   std::string());
+      }
+    }
+    absl::StrAppend(&description,
+                    FormatCommentKeyValueList(enum_comments, indent_level),
+                    "\n");
+  }
```
Comment (devbww) [inline generator/internal/discovery_type_vertex.cc:71]: I guess performance is not a concern, but continually re-evaluating `field["enum"]` at least smells a little.

File: generator/internal/discovery_type_vertex.cc [hunk 3]
```diff
@@ -44,6 +44,34 @@ absl::optional<std::string> CheckForScalarType(nlohmann::json const& j) {
...
+    std::vector<std::pair<std::string, std::string>> enum_comments;
+    for (unsigned int i = 0; i < field["enum"].size(); ++i) {
+      if (field.contains("enumDescriptions") &&
+          i < field["enumDescriptions"].size()) {
+        enum_comments.emplace_back(std::string(field["enum"][i]),
+                                   std::string(field["enumDescriptions"][i]));
+      } else {
+        enum_comments.emplace_back(std::string(field["enum"][i]),
+                                   std::string());
+      }
```
Comment (devbww) [inline generator/internal/discovery_type_vertex.cc:66]: This looks like an example of something I occasionally nit about, where much of the conditional behavior is not conditional.  That makes it harder to read.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @dbolduc and @devbww)

___
*[`generator/internal/discovery_type_vertex.cc` line 66 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#gh-1277027803:-NaS3gX04IMdtKHJOsid:bcaizm3) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d8c3ada07113b791f44a9e9517e3d0761d8cdeed/generator/internal/discovery_type_vertex.cc#L66)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

This looks like an example of something I occasionally nit about, where much of the conditional behavior is not conditional.  That makes it harder to read.
</blockquote></details>

Refactored the conditional bodies.
___
*[`generator/internal/discovery_type_vertex.cc` line 71 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#gh-1277025190:-NaS3eNw-S-s1F53gBex:b-5nqb5b) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d8c3ada07113b791f44a9e9517e3d0761d8cdeed/generator/internal/discovery_type_vertex.cc#L71)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

I guess performance is not a concern, but continually re-evaluating `field["enum"]` at least smells a little.
</blockquote></details>

Refactored to only call it once.
___
*[`generator/internal/discovery_type_vertex.cc` line 261 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#gh-1276905376:-NaS3_Lu-PiKtGnwwqc2:bjlgfer) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d8c3ada07113b791f44a9e9517e3d0761d8cdeed/generator/internal/discovery_type_vertex.cc#L261)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

nit: I cannot see why this is a class member vs. a standalone function. But surely the compiler would have yelled at you to make it `static`. I might just be blind.
</blockquote></details>

There's a call to this->package_name().

**Review by dbolduc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243)* status: 0 of 2 files reviewed, 3 unresolved discussions (waiting on @devbww and @scotthart)

___
*[`generator/internal/discovery_type_vertex.cc` line 261 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#gh-1276905376:-NaS5GbFBMDrAgbBKxjw:byism81) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d8c3ada07113b791f44a9e9517e3d0761d8cdeed/generator/internal/discovery_type_vertex.cc#L261)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

There's a call to this->package_name().
</blockquote></details>

Ah, I see it now! Thanks.

**Review by devbww — COMMENTED** [review]
> Reviewed 1 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243)* status: all files reviewed, 5 unresolved discussions (waiting on @scotthart)

___
*[`generator/internal/discovery_type_vertex.h` line 125 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaSGdOF4bSO6uqHlJnb:-NaSGdOF4bSO6uqHlJnc:b-568ero) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/33b38391fb25dcec2e1dd52ff383b7972991c5b3/generator/internal/discovery_type_vertex.h#L125)):*
> ```objc
>  private:
>   Status UpdateTypeNames(
>       std::map<std::string, DiscoveryTypeVertex> const& types,
> ```

It looks like we never included `<map>` (this not being the first use).
___
*[`generator/internal/discovery_type_vertex.cc` line 58 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaSIPGW5Oj71xDHNL6p:-NaSIPGW5Oj71xDHNL6q:b-yv6qoy) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/33b38391fb25dcec2e1dd52ff383b7972991c5b3/generator/internal/discovery_type_vertex.cc#L58)):*
> ```cpp
>   auto enum_field = field.value("enum", nlohmann::json{});
>   if (!enum_field.empty()) {
>     auto const& enum_field = field["enum"];
> ```

This looks a little weird.  I'm at least surprised that you don't get some kind of tidy warning about redefining a local variable.

Anyway, I know precious little about the API, but it seems that something like this (untested) might be better:
```
  auto enum_field = field.find("enum");
  if (enum_field != field.end() && !enum_field->empty()) {
```
___
*[`generator/internal/discovery_type_vertex.cc` line 60 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaSL8eD34EuN4J0UiWc:-NaSL8eD34EuN4J0UiWd:bly9twq) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/33b38391fb25dcec2e1dd52ff383b7972991c5b3/generator/internal/discovery_type_vertex.cc#L60)):*
> ```cpp
>     auto const& enum_field = field["enum"];
>     std::vector<std::pair<std::string, std::string>> enum_comments;
>     for (unsigned int i = 0; i < enum_field.size(); ++i) {
> ```

It looks like nlohmann has a `size_type` (== `std::size_t`) that would be preferable.
___
*[`generator/internal/discovery_type_vertex.cc` line 64 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaSLfZx67knUPjuvgIN:-NaSLfZx67knUPjuvgIO:b-id6d7d) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/33b38391fb25dcec2e1dd52ff383b7972991c5b3/generator/internal/discovery_type_vertex.cc#L64)):*
> ```cpp
>       if (field.contains("enumDescriptions") &&
>           i < field["enumDescriptions"].size()) {
>         enum_description = std::string(field["enumDescriptions"][i]);
> ```

`field["enumDescriptions"]` could be hoisted out of the loop.

**Review by devbww — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243)* status: all files reviewed, 6 unresolved discussions (waiting on @scotthart)

___
*[`generator/internal/discovery_type_vertex.cc` line 66 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaSP8aj0GeBpvjkDg7V:-NaSP8aj0GeBpvjkDg7W:bd3bpd9) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/33b38391fb25dcec2e1dd52ff383b7972991c5b3/generator/internal/discovery_type_vertex.cc#L66)):*
> ```cpp
>         enum_description = std::string(field["enumDescriptions"][i]);
>       }
>       enum_comments.emplace_back(std::string(enum_field[i]), enum_description);
> ```

Is there a `move` opportunity there?

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243)* status: all files reviewed, 3 unresolved discussions (waiting on @devbww)

___
*[`generator/internal/discovery_type_vertex.h` line 125 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaSGdOF4bSO6uqHlJnb:-NaSYqum0Tme9Ts-W5sb:b-6vjzb8) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/33b38391fb25dcec2e1dd52ff383b7972991c5b3/generator/internal/discovery_type_vertex.h#L125)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

It looks like we never included `<map>` (this not being the first use).
</blockquote></details>

Added to the header
___
*[`generator/internal/discovery_type_vertex.cc` line 58 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaSIPGW5Oj71xDHNL6p:-NaSYuODBh3RNu1oU_uX:bmle2yg) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/33b38391fb25dcec2e1dd52ff383b7972991c5b3/generator/internal/discovery_type_vertex.cc#L58)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

This looks a little weird.  I'm at least surprised that you don't get some kind of tidy warning about redefining a local variable.

Anyway, I know precious little about the API, but it seems that something like this (untested) might be better:
```
  auto enum_field = field.find("enum");
  if (enum_field != field.end() && !enum_field->empty()) {
```
</blockquote></details>

Removed the redefine as it was a leftover.
___
*[`generator/internal/discovery_type_vertex.cc` line 60 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaSL8eD34EuN4J0UiWc:-NaSZ-Jl4YHKgMYpbeDh:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/33b38391fb25dcec2e1dd52ff383b7972991c5b3/generator/internal/discovery_type_vertex.cc#L60)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

It looks like nlohmann has a `size_type` (== `std::size_t`) that would be preferable.
</blockquote></details>

Done.
___
*[`generator/internal/discovery_type_vertex.cc` line 64 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaSLfZx67knUPjuvgIN:-NaSZ0F402V_jQci5IBG:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/33b38391fb25dcec2e1dd52ff383b7972991c5b3/generator/internal/discovery_type_vertex.cc#L64)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

`field["enumDescriptions"]` could be hoisted out of the loop.
</blockquote></details>

Done.
___
*[`generator/internal/discovery_type_vertex.cc` line 66 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaSP8aj0GeBpvjkDg7V:-NaSZ1KH86b5NT0BsT1V:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/33b38391fb25dcec2e1dd52ff383b7972991c5b3/generator/internal/discovery_type_vertex.cc#L66)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Is there a `move` opportunity there?
</blockquote></details>

Done.

**Review by devbww — APPROVED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243)* status: all files reviewed, 4 unresolved discussions (waiting on @scotthart)

___
*[`generator/internal/discovery_type_vertex.cc` line 58 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaSiMZKAb7Y_yURmj2m:-NaSiMZKAb7Y_yURmj2n:bsa9t4q) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2345e6c2ee8cc61d215d04f5ddf82284741ab205/generator/internal/discovery_type_vertex.cc#L58)):*
> ```cpp
>   auto enum_field = field.value("enum", nlohmann::json{});
>   if (!enum_field.empty()) {
>     auto const& enum_descriptions =
> ```

Nit: I believe we prefer not to capture values by const ref.
___
*[`generator/internal/discovery_type_vertex.cc` line 61 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaShc8q-V0V7Id-Rh-f:-NaShc8q-V0V7Id-Rh-g:b-1n2sei) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2345e6c2ee8cc61d215d04f5ddf82284741ab205/generator/internal/discovery_type_vertex.cc#L61)):*
> ```cpp
>         field.value("enumDescriptions", nlohmann::json{});
>     std::vector<std::pair<std::string, std::string>> enum_comments;
>     std::string enum_description;
> ```

Nit: I don't think you won anything moving this out of the loop, and it cost you having to add a `clear()`.
___
*[`generator/internal/discovery_type_vertex.cc` line 63 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaShLG2AkOM3joX3jQo:-NaShLG2AkOM3joX3jQp:b-ngqrfl) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2345e6c2ee8cc61d215d04f5ddf82284741ab205/generator/internal/discovery_type_vertex.cc#L63)):*
> ```cpp
>     std::string enum_description;
>     for (nlohmann::json::size_type i = 0; i < enum_field.size(); ++i) {
>       if (!enum_descriptions.empty() && i < enum_descriptions.size()) {
> ```

It looks like the `!enum_descriptions.empty()` is redundant.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243)* status: 1 of 2 files reviewed, 4 unresolved discussions (waiting on @devbww)

___
*[`generator/internal/discovery_type_vertex.cc` line 58 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaSiMZKAb7Y_yURmj2m:-NaSkndr-YqN4RlLtZz0:b-a2ujfl) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2345e6c2ee8cc61d215d04f5ddf82284741ab205/generator/internal/discovery_type_vertex.cc#L58)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Nit: I believe we prefer not to capture values by const ref.
</blockquote></details>

Changed.
___
*[`generator/internal/discovery_type_vertex.cc` line 61 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaShc8q-V0V7Id-Rh-f:-NaSkpAX2DcnkEUbYKMG:b-qj4lsq) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2345e6c2ee8cc61d215d04f5ddf82284741ab205/generator/internal/discovery_type_vertex.cc#L61)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Nit: I don't think you won anything moving this out of the loop, and it cost you having to add a `clear()`.
</blockquote></details>

Reworked.
___
*[`generator/internal/discovery_type_vertex.cc` line 63 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaShLG2AkOM3joX3jQo:-NaSkrUs0-FlRdbeFD5-:ba3dfob) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2345e6c2ee8cc61d215d04f5ddf82284741ab205/generator/internal/discovery_type_vertex.cc#L63)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

It looks like the `!enum_descriptions.empty()` is redundant.
</blockquote></details>

Removed

**Review by devbww — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243)* status: 1 of 2 files reviewed, 2 unresolved discussions (waiting on @scotthart)

___
*[`generator/internal/discovery_type_vertex.cc` line 61 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12243#-NaShc8q-V0V7Id-Rh-f:-NaSmHIUDNfy7Yx_CbEw:b5jg5uj) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2345e6c2ee8cc61d215d04f5ddf82284741ab205/generator/internal/discovery_type_vertex.cc#L61)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Reworked.
</blockquote></details>

It looks like we jumped back to the non-conditional conditional shape, but OK.

**Pull request merged**